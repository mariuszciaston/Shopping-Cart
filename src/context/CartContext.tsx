import { CartContextType } from "@/types/types";
import { createContext, useContext } from "react";

export const CartContext = createContext<CartContextType | null>(null);

export const useCartContext = () => {
  const context = useContext(CartContext);
  if (!context) {
    throw new Error("useCartContext must be used within a CartProvider");
  }
  return context;
};
