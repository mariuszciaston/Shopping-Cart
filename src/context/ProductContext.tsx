import { ProductsContextType } from "@/types";
import { createContext, useContext } from "react";

export const ProductsContext = createContext<ProductsContextType | null>(null);

export const useProductsContext = () => {
  const context = useContext(ProductsContext);
  if (!context) {
    throw new Error(
      "useProductsContext must be used within a ProductsProvider",
    );
  }
  return context;
};
