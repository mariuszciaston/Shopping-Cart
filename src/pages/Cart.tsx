import { useCartContext } from "@/context/CartContext";
import { useProductsContext } from "@/context/ProductContext";
import Header from "@/components/layouts/Header";
import Main from "@/components/layouts/Main";
import Footer from "@/components/layouts/Footer";
import Wrapper from "@/components/ui/Wrapper";
import CartItem from "@/components/cart/CartItem";
import CartSummary from "@/components/cart/CartSummary";

const Cart = () => {
  const { productsData, error, loading } = useProductsContext();
  const { itemsInCart } = useCartContext();

  return (
    <Wrapper>
      <Header />
      <Main>
        <div className="flex flex-col gap-8 rounded-md bg-white p-4 shadow-md sm:p-8">
          {loading && <p>Loading...</p>}
          {error && <p>A network error was encountered</p>}

          {itemsInCart.length === 0 ? (
            <div>Cart is empty</div>
          ) : (
            productsData &&
            productsData
              .filter((product) =>
                itemsInCart.some((item) => item.id === product.id),
              )
              .map((product) => <CartItem key={product.id} product={product} />)
          )}

          <hr />

          <CartSummary productsData={productsData} itemsInCart={itemsInCart} />
        </div>
      </Main>
      <Footer />
    </Wrapper>
  );
};

export default Cart;
