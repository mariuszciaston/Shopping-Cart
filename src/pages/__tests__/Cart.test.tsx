import { render, screen } from "@testing-library/react";
import { BrowserRouter } from "react-router-dom";
import { beforeEach, describe, expect, it, vi, Mock } from "vitest";
import { useCartContext } from "@/context/CartContext";
import { useProductsContext } from "@/context/ProductContext";
import Cart from "../Cart";

vi.mock("@/context/CartContext", () => ({
  useCartContext: vi.fn(),
}));

vi.mock("@/context/ProductContext", () => ({
  useProductsContext: vi.fn(),
}));

const mockUseCartContext = useCartContext as Mock;
const mockUseProductsContext = useProductsContext as Mock;

const mockProducts = [
  {
    id: 1,
    title: "Product 1",
    price: 19.99,
    image: "image1.jpg",
    description: "Description 1",
  },
  {
    id: 2,
    title: "Product 2",
    price: 29.99,
    image: "image2.jpg",
    description: "Description 2",
  },
];

describe("Cart page component", () => {
  beforeEach(() => {
    mockUseProductsContext.mockReturnValue({
      productsData: mockProducts,
      error: null,
      loading: false,
    });
  });

  it("renders header, main, and footer components", () => {
    mockUseCartContext.mockReturnValue({
      itemsInCart: [],
    });

    render(
      <BrowserRouter>
        <Cart />
      </BrowserRouter>,
    );

    expect(screen.getByRole("banner")).toBeInTheDocument();
    expect(screen.getByRole("main")).toBeInTheDocument();
    expect(screen.getByRole("contentinfo")).toBeInTheDocument();
  });

  it("displays empty cart message when cart is empty", () => {
    mockUseCartContext.mockReturnValue({
      itemsInCart: [],
    });

    render(
      <BrowserRouter>
        <Cart />
      </BrowserRouter>,
    );

    expect(screen.getByText("Cart is empty")).toBeInTheDocument();
  });

  it("displays loading state", () => {
    mockUseCartContext.mockReturnValue({
      itemsInCart: [{ id: 1, quantity: 1 }],
    });
    mockUseProductsContext.mockReturnValue({
      productsData: null,
      error: null,
      loading: true,
    });

    render(
      <BrowserRouter>
        <Cart />
      </BrowserRouter>,
    );

    expect(screen.getByText("Loading...")).toBeInTheDocument();
  });

  it("displays error message when network error occurs", () => {
    mockUseCartContext.mockReturnValue({
      itemsInCart: [{ id: 1, quantity: 1 }],
    });
    mockUseProductsContext.mockReturnValue({
      productsData: null,
      error: "A network error was encountered",
      loading: false,
    });

    render(
      <BrowserRouter>
        <Cart />
      </BrowserRouter>,
    );

    expect(
      screen.getByText("A network error was encountered"),
    ).toBeInTheDocument();
  });

  it("renders cart items and summary when items exist in cart", () => {
    mockUseCartContext.mockReturnValue({
      itemsInCart: [
        { id: 1, quantity: 2 },
        { id: 2, quantity: 2 },
      ],
    });

    render(
      <BrowserRouter>
        <Cart />
      </BrowserRouter>,
    );

    expect(screen.getByText("Product 1")).toBeInTheDocument();
    expect(screen.getByText("Product 2")).toBeInTheDocument();
    expect(screen.getByTestId("total-amount")).toHaveTextContent("$99.96");
  });

  it("matches snapshot", () => {
    mockUseCartContext.mockReturnValue({
      itemsInCart: [{ id: 1, quantity: 1 }],
    });

    const { asFragment } = render(
      <BrowserRouter>
        <Cart />
      </BrowserRouter>,
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
