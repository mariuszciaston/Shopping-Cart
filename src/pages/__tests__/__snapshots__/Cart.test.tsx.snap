// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Cart page component > matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="flex h-[calc(100vh)] flex-col"
    id="wrapper"
  >
    <header
      class="mx-auto flex w-full max-w-[1280px] flex-wrap items-center justify-between gap-4 p-4 sm:gap-8 sm:p-8"
    >
      <h1
        class="text-3xl font-bold"
      >
        <a
          class="rounded-md border-2 border-black"
          data-discover="true"
          href="/"
        >
          <span
            class="bg-black text-white"
          >
             fake 
          </span>
          <span>
             STORE 
          </span>
        </a>
      </h1>
      <nav
        class="order-0 sm:order-0 md:order-2"
      >
        <ul
          class="flex items-center gap-8"
        >
          <li>
            <a
              data-discover="true"
              href="/"
            >
              Home
            </a>
          </li>
          <li>
            <a
              data-discover="true"
              href="/products"
            >
              Products
            </a>
          </li>
          <li>
            <a
              data-discover="true"
              data-testid="cart-link"
              href="/cart"
            >
              <div
                class="relative"
              >
                <svg
                  class="lucide lucide-shopping-basket h-8 w-8"
                  fill="none"
                  height="24"
                  stroke="currentColor"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                  viewBox="0 0 24 24"
                  width="24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="m15 11-1 9"
                  />
                  <path
                    d="m19 11-4-7"
                  />
                  <path
                    d="M2 11h20"
                  />
                  <path
                    d="m3.5 11 1.6 7.4a2 2 0 0 0 2 1.6h9.8a2 2 0 0 0 2-1.6l1.7-7.4"
                  />
                  <path
                    d="M4.5 15.5h15"
                  />
                  <path
                    d="m5 11 4-7"
                  />
                  <path
                    d="m9 11 1 9"
                  />
                </svg>
                <span
                  class="absolute -bottom-1 -right-1 flex h-5 min-w-5 items-center justify-center rounded-full bg-red-500 text-xs text-white"
                  data-testid="cart-item-count"
                >
                  1
                </span>
              </div>
            </a>
          </li>
        </ul>
      </nav>
    </header>
    <main
      class="flex-1 bg-gray-100"
    >
      <div
        class="mx-auto grid h-full w-full max-w-[1280px] content-start py-8 sm:p-8"
      >
        <div
          class="flex flex-col gap-8 rounded-md bg-white p-4 shadow-md sm:p-8"
        >
          <div
            class="flex flex-1 flex-wrap items-center justify-between gap-4"
          >
            <div
              class="flex items-center gap-4 sm:min-w-[480px] sm:flex-1"
            >
              <a
                data-discover="true"
                href="/products/1"
              >
                <img
                  alt="Product 1"
                  class="aspect-square w-32 min-w-32 rounded-md border bg-white object-contain p-4"
                  src="image1.jpg"
                />
              </a>
              <a
                data-discover="true"
                href="/products/1"
              >
                <p
                  class="flex flex-col"
                >
                  <span
                    class="font-bold"
                    data-testid="product-title"
                  >
                    Product 1
                  </span>
                  <span
                    data-testid="product-price"
                  >
                    $19.99
                  </span>
                </p>
              </a>
            </div>
            <div
              class="flex flex-1 items-center gap-4"
            >
              <div
                class="flex gap-2"
              >
                <button
                  class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 border border-input bg-background shadow-sm hover:bg-accent hover:text-accent-foreground h-9 px-4 py-2 text-lg"
                  data-testid="quantity-lower"
                >
                  -
                </button>
                <input
                  class="flex h-9 rounded-md border border-input bg-transparent px-3 py-1 text-base shadow-sm transition-colors file:border-0 file:bg-transparent file:text-sm file:font-medium file:text-foreground placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:cursor-not-allowed disabled:opacity-50 md:text-sm m-none w-[4rem] pr-0 text-center"
                  data-testid="quantity-input"
                  max="99"
                  min="1"
                  type="number"
                  value="1"
                />
                <button
                  class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 border border-input bg-background shadow-sm hover:bg-accent hover:text-accent-foreground h-9 px-4 py-2 text-lg"
                  data-testid="quantity-boost"
                >
                  +
                </button>
              </div>
              <svg
                class="lucide lucide-trash2 cursor-pointer hover:text-red-500"
                data-testid="remove-item"
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M3 6h18"
                />
                <path
                  d="M19 6v14c0 1-1 2-2 2H7c-1 0-2-1-2-2V6"
                />
                <path
                  d="M8 6V4c0-1 1-2 2-2h4c1 0 2 1 2 2v2"
                />
                <line
                  x1="10"
                  x2="10"
                  y1="11"
                  y2="17"
                />
                <line
                  x1="14"
                  x2="14"
                  y1="11"
                  y2="17"
                />
              </svg>
              <p
                class="flex flex-1 justify-end"
                data-testid="sum"
              >
                $19.99
              </p>
            </div>
          </div>
          <hr />
          <div
            class="flex justify-between"
          >
            <div
              class="text-lg font-bold"
            >
              Total  
            </div>
            <div
              class="text-lg font-bold"
              data-testid="total-amount"
            >
              $19.99
            </div>
          </div>
          <button
            class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 text-primary-foreground shadow h-9 mx-auto bg-blue-500 px-8 py-6 text-lg font-bold hover:bg-blue-500"
          >
            Checkout
          </button>
        </div>
      </div>
    </main>
    <footer
      class="mx-auto flex w-full max-w-[1280px] items-center justify-center p-4 text-sm sm:p-8"
    >
      <span>
        Mariusz Ciastoń 2025  
      </span>
      <a
        href="https://github.com/mariuszciaston/"
        rel="noopener"
        target="_blank"
      >
        <i
          aria-hidden="true"
          class="fab fa-github fa-2x transition-transform duration-500 ease-in-out hover:rotate-1turn hover:scale-125 hover:transform"
        />
      </a>
    </footer>
  </div>
</DocumentFragment>
`;
